docs/proposals/20200602-machine-deletion-phase-hooks.md

```diff
@@ -0,0 +1,301 @@
+* Watch machine objects and determine when an appropriate action must be taken.
+* After completing the desired hook action, remove the hook annotation.
+##### Owning Controllers may
+* Watch machine objects and add a hook annotation as desired by the operator
+* After completing a hook, set an addition annotation that indicates this
+controller has finished for dependency ordering, or perform some other action
+to signal other Owning Controllers that they may proceed.
+
+#### Determining when to take action
+
+An Owning Controller should watch machines and determine when is the
+best time to take action.
+
+For example, if an Owning Controller manages a lifecycle hook at the
+pre-drain lifecycle-point, then that controller should take action
+immediately after a machine has a DeletionTimestamp or enters the
+"Deleting" phase.
+
+Fine-tuned coordination is not possible at this time; eg, it's not
+possible to execute a pre-term hook only after a node has been drained.
+
+##### Failure Mode
+It is entirely up to the owning controller to determine when it is prudent to
+remove a particular lifecycle hook.  Some controllers may want to 'give up'
+after a certain time period, and others may want to block indefinitely.
+Cluster operators should consider the characteristics of each controller before
+utilizing them in their clusters.
+
+
+### Risks and Mitigations
+
+* Annotation keys must conform to length limits: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set
+* Requires well-behaved controllers and admins to keep things running
+smoothly.  Would be easy to disrupt machines with poor configuration.
+* Troubleshooting problems may increase in complexity, but this is
+mitigated mostly by the fact that these hooks are opt-in.  Operators
+will or should know they are consuming these hooks, but a future proliferation
+of the cluster-api could result in these components being bundled as a
+complete solution that operators just consume.  To this end, we should
+update any troubleshooting guides to check these hook points where possible.
+
+
+## Alternatives
+
+### Custom Machine Controller
+Require advanced users to fork and customize.  This can already be done if someone chooses, so not much of a solution.
+
+### Finalizers
+We define additional finalizers, but this really only implies the deletion lifecycle point.  A misbehaving controller that
+accidentally removes finalizers could have undesireable
+effects.
+
+### Status Field
+Harder for users to modify or set hooks during machine creation.  How would a user remove a hook if a controller that is supposed to remove it is misbehaving?  We’d probably need an annotation like ‘skip-hook-xyz’ or similar and that seems redundant to just using annotations in the first place
+
+### Spec Field
+We probably don’t want other controllers dynamically adding and removing spec fields on an object.  It’s not very declarative to utilize spec fields in that way.
```

Comment:
It might be good to go into this alternative in more detail. For example, if it would be common for the user to configure the lifecycle hooks when they create the Cluster, it would seem very natural to configure those hooks through the spec (as well as being more visible that they are configured) compared to annotations.

That said, I think for it to be effective it would have to be tied to the below section of being spec fields that are linked to a hook-specific CRD, which would then allow for the hook to provide status back to the Machine Controller (we could define required fields for a hook implementation similar to what we do for other "providers').

The combination above would also allow for more visibility to the user when things are wedged or in the case that a controller "gives up" the behavior could be for the Machine Controller to orphan the hook CRD resource so the failure state remains.

Reply:
I don't really feel the need to articulate in such fine detail something that's intentionally not being implemented.  I think my description is pretty succinct.  For one, it's not very declarative to have mutating fields in the spec.  For two, it's not very declarative to have other components add hooks to the spec at some later point in a machine's lifecycle.  Finally, this configuration is not related to the machine object itself, rather the behavior of the machine-controller.

> That said, I think for it to be effective it would have to be tied to the below section of being spec fields that are linked to a hook-specific CRD

That is one of several possibilities.  There could be a spec field without a CRD.  There could be a CRD without a spec field.  There could be spec field + status.  There could be CRD + status.  There could be CRD + Spec + Status.  There could be annotations + CRD.  There could be spec + annotations.  There could be annotations + status field.  There could be CRDs + annotations + spec + status.  There could be 1 CRD per hook or there could be 1 CRD per machine, or there could be 1 CRD that utilizes selectors.

Reply:
I'd like to explore the possibility of spec fields in a _**future**_ iteration of this feature. @detiber is it ok to mark this comment resolved?